src/utils/actions.js

```diff
@@ -0,0 +1,396 @@
+            // get value by path
+            const date = getByPath(resource, path); // assume its a datetime?
+            return compareDates(date, filterPeriod);
+          });
+          filteredResources.push.apply(filteredResources, dateFiltered);
+        } else {
+          // path or searchParam are required
+          const dateFiltered = resources.filter(resource => {
+            // get value by param
+            const date = getBySearchParam(resource, searchParam); // assume its a datetime?
+            return compareDates(date, filterPeriod);
+          });
+          filteredResources.push.apply(filteredResources, dateFiltered);
+        }
+      } else {
+        // if there's no date filter, we can just push all the resources
+        // in to be filtered by code
+        filteredResources.push.apply(filteredResources, resources);
+      }
+
+      const codeFilter = input.codeFilter;
+      if (codeFilter) {
+        // code filter isn't under the same constraint as datefilter
+        // we can assume that it's just a path
+        const path = codefilter.path;
+        const searchParam = codeFilter.searchParam;
+        if (path) {
+          // no need to refilter resources
+          const filteredResources = filteredResources.filter(resource => {
+            const code = getByPath(resource, path);
+            return compareCodes(code, codeFilter.code); // codeFilter.code is of type Coding, could also be a valueSet?
+          });
+        } else if (searchParam) {
+          const filteredResources = filteredResources.filter(resource => {
+            const code = getBySearchParam(resource, searchParam);
+            return compareCodes(code, codeFilter.code);
+          });
+        }
+      }
+      return context.records.push(filteredResources);
+    });
+  }
+}
+function executeResearchQuery(context) {
+  // TODO: what's a datamart?
+}
+function getByPath(resource, path) {
+  // TODO: function to search resources by path
+  // and return correct attribute
+  return resource;
+}
+
+function compareCodes(code1, code2) {
+  // assume they're both coding
+  return code1.code === code2.code && code1.system === code2.system;
+}
+
+function getBySearchParam(resource, searchParam) {
+  // TODO: function to search resources by param
+  // and return correct attribute
+  return resource.date;
+}
+
+function compareDates(period1, period2) {
+  // TODO: returns true if the periods overlap
+}
+function createBundle(records) {
+  const bundle = {
+    resourceType: 'Bundle',
+    type: 'collection',
+    entry: []
+  };
+
+  records.forEach(record => {
+    bundle.push({
+      resource: record
+    });
+  });
+
+  return bundle;
+}
+
+function getPatient(context) {
+  // assume the patient is only an ID
+  const patientId = context.patient;
+  const patient = context.client.source.read(patientId);
+  context.patient = patient;
+  return context;
+}
+
+function evaluateExpression(expression, resource) {
```

Comment:
Perhaps I'm making this up but I thought I remember Rob saying MedMorph was trying to go fully fhir path and include no cql

Reply:
Yeah the latest was that the spec allows both but requires FHIRPath. So at that point there's really no reason to also include CQL. But again the spec does allow it